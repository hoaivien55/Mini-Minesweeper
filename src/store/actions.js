const FETCH_DATA = "FETCH_DATA";
const CHANGE_GAME_LEVEL = "CHANGE_GAME_LEVEL";

export {
    FETCH_DATA,
    CHANGE_GAME_LEVEL
}