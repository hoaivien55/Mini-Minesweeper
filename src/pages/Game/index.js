import { Fragment, memo, useContext, useEffect, useState, useRef } from "react";
import { useNavigate } from "react-router-dom";

import { fetchMines } from "../../api";
import { FETCH_DATA } from "../../store/actions";
import Popup from "../../components/Popup";
import { Container, Row, Error } from "../../components/styles";
import { Store } from "../../store";
import GameBox from "./components/GameBox";
import Header from "./components/Header";
import Loader from "../../components/Loader";
import { getLevelSelectedObj } from "../../utils";

const Game = () => {
	const navigate = useNavigate();
	const { appState, dispatch } = useContext(Store);
	const [_gameStatus, _setGameStatus] = useState({
		end: false,
		begin: false,
		count: 1,
		timer: false,
		gameTime: "",
	});
	const [_loader, _setLoader] = useState(false);
	const [_error, _setError] = useState("");

	const { rawMines, levels, levelSelected } = appState;
	const level = getLevelSelectedObj(levels, levelSelected);
	const { size } = level;
	const timerRef = useRef();

	useEffect(() => {
		async function _fetchMines() {
			if (rawMines.length === 0) {
				// default mines
				// console.log("reload mines")
				_setError("");
				_setLoader(true);
				const { mines, size } = level;
				const data = await fetchMines({ mines, size });
				if (data && data.length > 0) {
					dispatch({
						type: FETCH_DATA,
						payload: data,
					});
				} else {
					_setError("Cannot call api. Please try again.");
				}
				_setLoader(false);
			}
		}
		_fetchMines();

		// eslint-disable-next-line
	}, []);

	useEffect(() => {
		// console.log("change raw mines")
		_setGameStatus({
			count: _gameStatus.count + 1,
			end: false,
			timer: false,
		});
		// eslint-disable-next-line
	}, [rawMines]);

	const _startGame = () => {
		_setGameStatus({
			..._gameStatus,
			timer: true,
			gameTime: "",
		});
	};

	const _endGame = (win) => {
		// console.log("endgame", win);
		const gameTime = timerRef.current.getTime();
		_setGameStatus({
			..._gameStatus,
			timer: false,
			end: true,
			win,
			gameTime,
		});
	};

	const _getLevelName = () => {
		const level = getLevelSelectedObj(levels, levelSelected);
		// console.log("level", level);
		return (level && level.name) || "";
	};

	const _popupMessage = () => {
		return `You ${(_gameStatus.win && "won") || "lost"} the game in ${
			_gameStatus.gameTime || "00:00:00"
		}`;
	};

	const _goHomepage = () => {
		navigate("/");
	};

	const _newGame = async () => {
		// console.log("restart game");
		_setError("");
		const { mines, size } = level;
		const data = await fetchMines({ mines, size });
		if (data && data.length > 0) {
			dispatch({
				type: FETCH_DATA,
				payload: data,
			});
		} else {
			_setError("Cannot call api. Please try again.");
		}
		_setGameStatus({
			count: _gameStatus.count + 1,
			end: false,
			timer: false,
			gameTime: "",
		});
		_setLoader(false);
	};

	const _popup = (message) => {
		return (
			_gameStatus.end && (
				<Popup
					message={message || _popupMessage()}
					acceptLabel={"new game"}
					rejectLabel={"home page"}
					reject={_goHomepage}
					accept={_newGame}
				/>
			)
		);
	};

	return (
		<Fragment>
			{_popup()}
			<Container backgroundColor="#ffffff" key={_gameStatus.count}>
				<Loader show={_loader} />
				<Header
					timer={_gameStatus.timer}
					level={_getLevelName()}
					timerRef={timerRef}
				/>
				<GameBox
					size={size}
					mines={rawMines}
					endGame={_endGame}
					startGame={_startGame}
				/>
				<Row>
					<Error>{_error}</Error>
				</Row>
			</Container>
		</Fragment>
	);
};

export default memo(Game);
