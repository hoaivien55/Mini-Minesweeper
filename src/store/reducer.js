import { cacheGameInfo } from "../utils";
import { FETCH_DATA, CHANGE_GAME_LEVEL } from "./actions";

function reducer(state, action) {
	switch (action.type) {
		case FETCH_DATA:
			// console.log(FETCH_DATA);
			return { ...state, rawMines: action.payload };
		case CHANGE_GAME_LEVEL:
			// console.log(CHANGE_GAME_LEVEL);
			cacheGameInfo(action.payload);
			return { ...state, ...action.payload };
		default:
			return state;
	}
}

export default reducer;
