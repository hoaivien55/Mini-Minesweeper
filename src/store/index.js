import React, { createContext, useReducer } from "react";
// import { testData } from "../utils";

import reducer from "./reducer";
export const Store = createContext("");

const defaultState = {
	userName: "Adam",
	levelSelected: null,
	matrix: [],
	rawMines: [],
	levels: [
		{
			size: 9,
			mines: 10,
			id: 1,
			name: "Beginner",
		},
		{
			size: 16,
			mines: 40,
			id: 2,
			name: "Advantage",
		},
	],
};

export function StoreProvider({ children, initialState = defaultState }) {
	const [appState, dispatch] = useReducer(reducer, initialState);
	const value = { appState, dispatch };
	// console.log(appState);
	return <Store.Provider value={value}>{children}</Store.Provider>;
}
