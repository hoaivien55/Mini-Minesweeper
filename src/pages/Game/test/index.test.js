import {
	render,
	screen,
	act,
	waitFor,
	fireEvent,
} from "@testing-library/react";
import renderer from "react-test-renderer";

import { BrowserRouter } from "react-router-dom";
import Game from "..";
import { StoreProvider } from "../../../store";
import { mines9x9, defaultState } from "./data";

// mock react router: useNavigate
const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
	...jest.requireActual("react-router-dom"),
	useNavigate: () => mockedUsedNavigate,
}));

test("Game: Virtual DOM snapshot", () => {
	const tree = renderer
		.create(
			<BrowserRouter>
				<StoreProvider initialState={defaultState}>
					<Game />
				</StoreProvider>
			</BrowserRouter>
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});

test("Game: renders page", async () => {
	global.fetch = jest.fn(() =>
		Promise.resolve({
			json: () =>
				Promise.resolve({
					msg: "success",
					data: mines9x9,
				}),
		})
	);
	act(() => {
		render(
			<BrowserRouter>
				<StoreProvider initialState={defaultState}>
					<Game />
				</StoreProvider>
			</BrowserRouter>
		);
	});
	await waitFor(() => {
		const boxs = document.querySelectorAll(".close");
		expect(boxs.length).toBe(81);
	});
});


test("Game: click boom", async () => {
	act(() => {
		render(
			<BrowserRouter>
				<StoreProvider initialState={defaultState}>
					<Game />
				</StoreProvider>
			</BrowserRouter>
		);
	});
    jest.clearAllMocks();
	await waitFor(() => {
		const boxs = document.querySelectorAll(".close");
		fireEvent.click(boxs[2])
        expect(screen.getByText(/You lost/)).toBeInTheDocument()
	});
});

// .....