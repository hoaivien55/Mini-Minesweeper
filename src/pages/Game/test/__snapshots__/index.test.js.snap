// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game: Virtual DOM snapshot 1`] = `
.c0 {
  position: absolute;
  left: 50%;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  margin: 0 auto;
  width: auto;
  min-width: 300px;
  min-height: 200px;
  padding: 20px;
  background-color: #ffffff;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0px;
}

.c3 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  color: red;
}

.c5 {
  border: 3px solid;
  border-color: #f7e8ae #bab7a9 #bab7a9 #dbd5b9;
  min-width: 30px;
  height: 30px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-weight: bold;
  font-size: 17px;
}

.c5.flag::before {
  content: "üö©";
  position: absolute;
  width: 100%;
  height: 100%;
  left: 50%;
  top: 50%;
  -webkit-transform: translate(-50%,-30%);
  -ms-transform: translate(-50%,-30%);
  transform: translate(-50%,-30%);
}

.c5.boom {
  background-color: #cecab7;
  border-color: #9c998d;
}

.c5.boom::before {
  content: "üí£";
  position: absolute;
  width: 100%;
  height: 100%;
  left: 50%;
  top: 50%;
  -webkit-transform: translate(-50%,-30%);
  -ms-transform: translate(-50%,-30%);
  transform: translate(-50%,-30%);
}

.c5.empty,
.c5.num {
  background-color: #cecab7;
  border-color: #9c998d;
  color: black;
}

.c4 {
  display: grid;
  grid-template-columns: repeat(9,1fr);
  border: 10px solid #a5a08d;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2 {
  font-size: 1.1em;
}

.c2 > div:not(:first-child) {
  border-left: 1px solid #80808091;
}

<div
  className="c0"
>
  <div
    className="c1 c2"
  >
    <div
      className="c3"
    >
      Level: 
      Beginner
    </div>
    <div
      className="c3"
    >
      00:00:00
    </div>
    <div
      className="c3"
      title="Right-click an close square to flag it and no limit flag"
    >
      ‚ùì
    </div>
  </div>
  <div
    className="c4"
  >
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
    <div
      className="c5 close"
      color=""
      onClick={[Function]}
      onContextMenu={[Function]}
    />
  </div>
  <div
    className="c1"
  >
    <div
      className="c6"
    />
  </div>
</div>
`;
